program.c

```diff
@@ -172,3 +172,26 @@ int program_find_bootable_partition(void)
 
 	return part;
 }
+
+void program_unload(void)
+{
+	struct program *program;
+
+	programes_last = NULL;
+
+	for (program = programes; program; program = program->next) {
+		if (program->filename)
+			xmlFree(program->filename);
+		if (program->start_sector)
+			xmlFree(program->start_sector);
+		if (program->label)
+			xmlFree(program->label);
+
+		if (programes_last)
+			free(programes_last);
+
+		programes_last = program;
```

Comment:
Use a second variable to track the "next" element as you free up the current "program".

for (program = programs; program; program = next) {
   next = program->next;
   free(program);
}